056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,131 @@
+in a pending state.
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to roll nodes
+Roll in order:
+- Unready controller/combined nodes
+- Ready controller/combined nodes
+- Active controller
+- Unready broker-only nodes
+- Ready broker-only nodes
+
+
+### Triggers
+The following are some of the triggers that would roll a KRaft controller or combined node:
+- Pod is annotated for manual rolling update
+- Pod is unready because it is in `CrashLoopBackOff`, `ImagePullBackOff`, `ContainerCreating` or `Pending` and `Unschedulable`, 
+- Controller's configuration has changed and cannot be updated dynamically.
+
+The triggers for broker remain the same as Zookeeper mode.
+
+**NOTE** There is a related proposal describing how to diff controller configuration currently being discussed: [PR #82](https://github.com/strimzi/proposals/pull/82)
+
+### Rollability checks
+
+The checks made by the KafkaRoller in different modes is described below.
+The checks include a new check for controllers to verify that rolling the node does not affect the quorum health.
+The proposed check is:
+- Fetch `lastCaughtUpTimestamp` for every controller node. `lastCaughtUpTimestamp` is the last millisecond timestamp at which a replica controller was known to be caught up with the quorum leader
+- Identify the `lastCaughtUpTimestamp` of the node that is the quorum leader .
+- Retrieve the current value of Kafka configuration `controller.quorum.fetch.timeout.ms`
+- Mark a node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`
+- Count each controller node that is caught up, including the leader (`numOfCaughtUpControllers`)
+- Can roll if: `numOfCaughtUpControllers > ceil(double) totalNumOfControllers / 2`
+
+In order to perform quorum check for controller, KafkaRoller creates admin client connection to the brokers. This is required until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) is implemented. If KafkaRoller cannot connect to any of the brokers, reconcilation for the controllers will be marked as failed due to not being able to determine the quorum health. KafkaRoller would then try to reconcile the brokers, which may help restoring admin client connections to them. The controllers quorum check will be retried in the next round of reconcilation.
```

Comment:
`will be retried in the next round of reconcilation` -> does that mean you will re-queue them and try them again after checking the brokers? Or that you will leave it to next reconciliation? Leaving it to the next recocniliation might possible leave it broken forever. But I'm not sure if it is something we can fix at the same time 🤔

Reply:
I also think that we should fail the reconciliation after a number of failed retries, in order to inform the user that there is some serious issue going on.

Reply:
So we can return `UnforceableProblem` to the scheduler, which would re-queue, delay and retry the controllers until maxAttempt is reached (in the meantime, the connections to brokers could also get resolved?). However we would not check the brokers until KafkaRoller finished reconciling the controllers.  This is to get the ordering correct (so that we can re-queue the active controller to the end of controller list, not the entire list including brokers). Once brokers are checked and attempted for rolling, the reconciliation will complete, we would not check the controllers again. In the next round of reconciliation, the controllers will be retried.  

@fvaleri yes, we already do that. Reconciliation fails if any of the pods has reached its max retry attempts or reported any other failure.

Reply:
@fvaleri @scholzj are you happy with Tina's responses on this thread? Can it be marked as resolved?